base/array.jl

```diff
@@ -2331,20 +2331,46 @@ Int64[]
 function findall(A)
     collect(first(p) for p in pairs(A) if last(p))
 end
+
 # Allocating result upfront is faster (possible only when collection can be iterated twice)
-function findall(A::AbstractArray{Bool})
-    n = count(A)
+function findall(f::Function, A::AbstractArray{Bool})
+    # Compute f for true and false only once
+    ft, ff = f(true), f(false)
```

Comment:
I think it is fine to assume a pure function here. I guess it could be argued that it is a bit wasteful to call `f(true)` on `findall(expensive_function, [false])`. I don't know if it is worth complicating the code for it though, probably not.

Reply:
That's true. On the other hand, it's much cheaper if the function takes any significant time to compute, since here there will only be 2 calls to f, instead of 1 call per element of A. I suppose the Base solution can only be the most generally applicable one, with special cases having to be handled by hand-crafted for loops.

Reply:
Yes, but in my example above there is only one element in the array.